nservicebus/messaging/messages-as-interfaces.md

```diff
@@ -1,21 +1,97 @@
 
-Imagine that the business logic represents a state machine with states X and Y. When the system gets into state X, it publishes the message `EnteredStateX`. When the system gets into state Y, it publishes the message `EnteredStateY`. (For more information on how to publish a message, see below.)
+## Using NServiceBus for dynamic dispatch
 
-In the next version of the system, a new state Z is added, which represents the co-existence of both X and Y states. This can be achieved by defining a message `EnteredStateZ` which inherits from `EnteredStateX` and `EnteredStateY`.
+Consider the following handler stubs for a `UserCreated` message which inherits from the `IEvent` stub interface:
 
-When the system publishes the `EnteredStateZ` event, the clients that are subscribed to either `EnteredStateX` or `EnteredStateY` are notified.
+```csharp
+public class SaveUser : IHandleMessages<UserCreated>
+{
+    public async Task Handle(UserCreated message, IMessageHandlerContext) {
+        // Do stuff when the user is created
+    }
+}
 
-Without the ability to inherit a message from other message types, composition techniques would be required, thereby preventing the infrastructure from automatically routing messages to pre-existing subscribers of the composed messages.
+public class Audit : IHandleMessages<IMessage>
+{
+    public async Task Handle(IMessage message, IMessageHandlerContext context) {
+        // Audit the message
+    }
+}
+```
+
+When a message of type `UserCreated` is published, both handlers will be invoked independently which in itself is useful. However, this idea becomes more powerful as the system evolves and more functionality is added.
+
+Assume there is a separate process that needs to happen if a user is created as part of a campaign. For example, say the subscriber must register statistics about the campaign. In NServiceBus, this can be done by defining a new event and handler:
+
+```csharp
+public class UserCreatedFromCampaign : UserCreated
+{
+    public Guid CampaignId { get; set;}
+}
+
+public class RecordStatistics : IHandleMessages<UserCreatedFromCampaign>
+{
+    public async Task Handle(UserCreatedFromCampaign, IMessageHandlerContext context) {
+        // Record statistics
+    }
+}
+```
+
+Here, when a `UserCreatedFromCampaign` message is published, the `RecordStatistics` handler is invoked as expected. But since `UserCreatedFromCampaign` inherits from `UserCreated`, the `SaveUser` handler is also invoked, along with the `Audit` handler (since `UserCreated` implements the `IMessage` stub interface). That is, new functionality has been added without having to change the existing handlers.
```

Comment:
```suggestion
Here, when a `UserCreatedFromCampaign` message is published, the `RecordStatistics` handler is invoked as expected. But since `UserCreatedFromCampaign` inherits from `UserCreated`, the `SaveUser` handler is also invoked, along with the `Audit` handler (since `UserCreated` implements the `IMessage` stub interface). 

The existing handlers remain untouched, and the entirely new functionality is added with just a few lines of new code. This is the beauty of dynamic dispatch.
```